gps/solution.go

```diff
@@ -58,7 +72,9 @@ const concurrentWriters = 16
 //
 // It requires a SourceManager to do the work. Prune options are read from the
 // passed manifest.
-func WriteDepTree(basedir string, l Lock, sm SourceManager, co CascadingPruneOptions, logger *log.Logger) error {
+//
+// If onWrite is not nil, it will be called after each project write. Calls are ordered and atomic.
+func WriteDepTree(basedir string, l Lock, sm SourceManager, co CascadingPruneOptions, onWrite func(WriteProgress)) error {
```

Comment:
Rather than log or nothing, callers get a rich type they can inspect and which also knows how to log itself.